File: python/hpc_instagram_geoprocessing.py
'''
File    : hpc_instagram_geoprocessing.py
Title   : HPC Instagram Geoprocessing
Author  : Ivan Ken Weng Chee 736901
Created : 08/04/2018
Purpose : COMP90024 2018S1 Assignment 1
'''

from mpi4py import MPI
import json
import numpy as np
import operator
import sys
import time

''' Obtain coordinates from Gridfile '''
def getCoordinates(data):
    coords = []
    for i in range(len(data['features'])):
        coords.append({})
        coords[i]['id'] = str(data['features'][i]['properties']['id'])
        coords[i]['xmin'] = data['features'][i]['properties']['xmin']
        coords[i]['xmax'] = data['features'][i]['properties']['xmax']
        coords[i]['ymin'] = data['features'][i]['properties']['ymin']
        coords[i]['ymax'] = data['features'][i]['properties']['ymax']
    return coords

''' Checks if a point is in a grid '''
def inside(location, coords):
    for i in range(len(coords)):
        if location['x'] >= coords[i]['xmin'] and location['x'] <= coords[i]['xmax'] and location['y'] >= coords[i]['ymin'] and location['y'] <= coords[i]['ymax']:
            return coords[i]['id']
    return None

''' Sorts and prints the final output '''
def order(posts, rows, cols):
    # Sorts the post counts
    sortedPosts = sorted(posts.items(), key=operator.itemgetter(1), reverse=True)
    sortedRows = sorted(rows.items(), key=operator.itemgetter(1), reverse=True)
    sortedCols = sorted(cols.items(), key=operator.itemgetter(1), reverse=True)

    # Final output
    print('// Rank by Unit')
    for i in range(len(sortedPosts)):
        print(sortedPosts[i][0] + ': ' + str(sortedPosts[i][1]) + ' posts')
    print('// Rank by Row')
    for i in range(len(sortedRows)):
        print(sortedRows[i][0] + '-Row: ' + str(sortedRows[i][1]) + ' posts')
    print('// Rank by Column')
    for i in range(len(sortedCols)):
        print('Column ' + sortedCols[i][0] + ': ' + str(sortedCols[i][1]) + ' posts')

''' Custom operation to reduce dictionary post counts '''
def counterSummation(x, y, datatype):
    for key in y:
        # Sum both values
        if key in x:
            x[key] += y[key]
        # Copy value over
        else:
            x[key] = y[key]
    return x

''' Main function '''
def main():

    # Checks for correct number of arguments
    if len(sys.argv) != 3:
        sys.exit()

    # Initialises MPI variables
    comm = MPI.COMM_WORLD
    size = comm.Get_size() # Number of processes
    rank = comm.Get_rank() # Process number
    name = MPI.Get_processor_name()

    # Master
    if rank == 0:
        # Starts timer
        start = time.clock()
        posts = {}
        rows = {}
        cols = {}
        coordspath = sys.argv[1]
        postspath = sys.argv[2]

        # Open coordinates
        with open(coordspath, 'r') as file:
            data = json.load(file)

        # Send coordinates to workers
        coords = getCoordinates(data)
        for i in range(1, size):
            comm.send(coords, dest=i)

        # Builds posts, rows, and columns lists
        for i in range(len(coords)):
            if coords[i]['id'] not in posts:
                posts[coords[i]['id']] = 0
            if coords[i]['id'][0] not in rows:
                rows[coords[i]['id'][0]] = 0
            if coords[i]['id'][1] not in cols:
                cols[coords[i]['id'][1]] = 0

        # Open posts
        with open(postspath, 'r') as file:

            # Multi core
            if (size > 1):
                turn = 1
                # Skips header and processes file
                next(file)
                for line in file:
                    try:
                        # Converts line into json format
                        post = json.loads(line.rstrip(", \r\n"))

                        # Finds and sends point coordinates to workers
                        if 'coordinates' in post['doc']:
                            loc = {
                                'x': post['doc']['coordinates']['coordinates'][1],
                                'y': post['doc']['coordinates']['coordinates'][0]
                            }
                            comm.send(loc, dest=turn)

                        # Change which worker to send to
                        if turn < (size - 1):
                            turn += 1
                        else:
                            turn = 1

                    except ValueError:
                        continue

                # Sends terminate signal to workers
                for i in range(1, size):
                    comm.send('terminate', dest=i)

            # Single core
            else:
                # Same process as multicore excluding MPI calls
                next(file)
                for line in file:
                    try:
                        post = json.loads(line.rstrip(", \r\n"))
                        if 'coordinates' in post['doc']:
                            loc = {
                                'x': post['doc']['coordinates']['coordinates'][1],
                                'y': post['doc']['coordinates']['coordinates'][0]
                            }
                            square = inside(loc, coords)
                            if square != None:
                                posts[square] += 1
                                rows[square[0]] += 1
                                cols[square[1]] += 1
                    except ValueError:
                        break

    # Worker
    elif rank > 0:
        # Receives coordinates from master
        coords = comm.recv(source=0)
        posts = {}
        rows = {}
        cols = {}

        # Builds posts, rows, and columns lists
        for i in range(len(coords)):
            if coords[i]['id'] not in posts:
                posts[coords[i]['id']] = 0
            if coords[i]['id'][0] not in rows:
                rows[coords[i]['id'][0]] = 0
            if coords[i]['id'][1] not in cols:
                cols[coords[i]['id'][1]] = 0

        # Keeps receiving lines from master until terminate signal is received
        while True:
            loc = comm.recv(source=0)
            if loc == 'terminate':
                break

            # Checks if points are in a grid and increment counters
            square = inside(loc, coords)
            if square != None:
                posts[square] += 1
                rows[square[0]] += 1
                cols[square[1]] += 1

    # Ensures processes are synchronised before combining results
    comm.barrier()

    # Reduces dictionary counts
    posts = comm.reduce(posts, root=0, op=MPI.Op.Create(counterSummation, commute=True))
    rows = comm.reduce(rows, root=0, op=MPI.Op.Create(counterSummation, commute=True))
    cols = comm.reduce(cols, root=0, op=MPI.Op.Create(counterSummation, commute=True))

    # Master prints final output
    if rank == 0:
        order(posts, rows, cols)

        # Prints the time taken
        end = time.clock()
        print('Time: ' + str(end - start) + 's')

if __name__ == '__main__':
    main()

File: java/HPCInstagramGeoprocessing.java
/*
 * File    : HPCInstagramGeoprocessing.java
 * Title   : HPC Instagram Geoprocessing
 * Author  : Ivan Ken Weng Chee 736901
 * Created : 10/04/2018
 * Purpose : COMP90024 2018S1 Assignment 1
 */

import java.io.*;
import java.util.*;
import mpi.*;

/* Processes a geocoded Instagram dataset and ranks results according to location */
public class HPCInstagramGeoprocessing {

  /* Main function */
  public static void main(String[] args) throws Exception {

    // Initialises MPI and other variables
    MPI.Init(args);
    int rank = MPI.COMM_WORLD.Rank();
    int size = MPI.COMM_WORLD.Size();
    int master = 0;
    int tag = 100;
    long startTime = 0;
    long endTime = 0;

    Boolean features = false;
    String gridfile;
    String instafile;
    String line;
    String header;
    BufferedReader br;

    Object[] postsArray = new Object[1];
    Object[] rowsArray = new Object[1];
    Object[] columnsArray = new Object[1];

    HashMap<String, Integer> posts = new HashMap<String, Integer>();
    HashMap<String, Integer> rows = new HashMap<String, Integer>();
    HashMap<String, Integer> columns = new HashMap<String, Integer>();

    ArrayList<Grid> grids;

    // Master
    if (rank == master) {

      // Checks for correct number of arguments
      if (args.length != 5) {
        System.exit(1);
      }

      gridfile = args[args.length - 2];
      instafile = args[args.length - 1];

      // Starts timer
      startTime = System.nanoTime();
      postsArray[0] = posts;
      rowsArray[0] = rows;
      columnsArray[0] = columns;
      grids = new ArrayList<Grid>();
      br = new BufferedReader(new FileReader(gridfile));

      // Processes the Instagram file
      try {
        while ((line = br.readLine()) != null) {
          if (line.equals("\"features\": [")) {
            features = true;
            continue;
          } else if (line.equals("]")) {
            features = false;
          }

          // A line containing grid information is found
          if (features) {
            String[] words = line.split(" ");
            Grid grid = new Grid();

            // Converts line into Grid objects
            for (int i = 0; i < words.length; i ++) {
              if (words[i].equals("\"id\":")) {
                grid.name = words[i + 1].replaceAll("\"", "").replaceAll(",", "");
                grid.row = grid.name.substring(0, 1);
                grid.column = grid.name.substring(1, 2);
              } else if (words[i].equals("\"xmin\":")) {
                grid.xmin = Float.parseFloat(words[i + 1].replaceAll(",", ""));
              } else if (words[i].equals("\"xmax\":")) {
                grid.xmax = Float.parseFloat(words[i + 1].replaceAll(",", ""));
              } else if (words[i].equals("\"ymin\":")) {
                grid.ymin = Float.parseFloat(words[i + 1].replaceAll(",", ""));
              } else if (words[i].equals("\"ymax\":")) {
                grid.ymax = Float.parseFloat(words[i + 1].replaceAll(",", ""));
              }
            }
            grids.add(grid);
          }
        }
      } catch (IOException e) {
        e.printStackTrace();
      }

      // Multi core
      if (size > 1) {

        // Send grid coordinates to workers
        Object[] sendObjectArray = new Object[1];
        sendObjectArray[0] = grids.toArray();
        for (int i = 1; i < size; i ++) {
          MPI.COMM_WORLD.Send(sendObjectArray, 0, 1, MPI.OBJECT, i, tag);
        }

        int worker = 1;
        br = new BufferedReader(new FileReader(instafile));

        // Processes the Instagram file
        try {
          header = br.readLine(); //First row
          while ((line = br.readLine()) != null) {

            // Send terminating signal to workers
            if (line.equals("]}")) {
              Object[] terminateObjectArray = new Object[1];
              terminateObjectArray[0] = (Object) "terminate";
              for (int i = 1; i < size; i ++) {
                MPI.COMM_WORLD.Send(terminateObjectArray, 0, 1, MPI.OBJECT, i, tag);
              }
              break;
            }

            // Send lines to workers alternatingly
            sendObjectArray[0] = (Object) line;
            MPI.COMM_WORLD.Send(sendObjectArray, 0, 1, MPI.OBJECT, worker, tag);

            // Switches worker to send to
            if (worker < (size - 1)) {
              worker += 1;
            } else {
              worker = 1;
            }
          }
        } catch (IOException e) {
          e.printStackTrace();
        }
      }

      // Single core
      else {

        // Builds grids, rows, and columns HashMaps
        for (Grid grid: grids) {
          posts.put(grid.name, 0);
          if (!rows.containsKey(grid.row)) {
            rows.put(grid.row, 0);
          }
          if (!columns.containsKey(grid.column)) {
            columns.put(grid.column, 0);
          }
        }

        br = new BufferedReader(new FileReader(instafile));

        // Processes the Instagram file
        try {
          header = br.readLine(); //First row
          while ((line = br.readLine()) != null) {
            // Stop reading
            if (line.equals("]}")) {
              break;
            }
            Integer temp = 0;
            String[] words = line.split(":");

            // Searches for point coordinates in each line
            for (int i = 0; i < words.length; i ++) {
              if (words[i].equals("\"Point\",\"coordinates\"")) {
                String[] latlon = words[i + 1].replaceAll("]}}},", "").replace("[", "").split(",");
                try {
                  Float x = Float.parseFloat(latlon[1]);
                  Float y = Float.parseFloat(latlon[0]);

                  // Checks if points are in a grid and increment counters
                  for (Grid grid: grids) {
                    if (grid.inGrid(x, y)) {
                      temp = posts.get(grid.name);
                      posts.remove(grid.name);
                      posts.put(grid.name, temp + 1);

                      temp = rows.get(grid.row);
                      rows.remove(grid.row);
                      rows.put(grid.row, temp + 1);

                      temp = columns.get(grid.column);
                      columns.remove(grid.column);
                      columns.put(grid.column, temp + 1);
                      break;
                    }
                  }
                } catch (NumberFormatException e) {
                  break;
                }
              }
            }
          }
        } catch (IOException e) {
          e.printStackTrace();
        }
      }
    }

    // Worker
    else {

      // Receive coordinates from master
      Object[][] recvGridsArray = new Object[1][];
      grids = new ArrayList<Grid>();
      MPI.COMM_WORLD.Recv(recvGridsArray, 0, 1, MPI.OBJECT, master, tag);
      for (int i = 0; i < recvGridsArray[0].length; i ++) {
        grids.add((Grid) recvGridsArray[0][i]);
      }

      // Receive lines from master
      Object[] recvObjectArray = new Object[1];
      line = null;

      // Builds grids, rows, and columns HashMaps
      for (Grid grid: grids) {
        posts.put(grid.name, 0);
        if (!rows.containsKey(grid.row)) {
          rows.put(grid.row, 0);
        }
        if (!columns.containsKey(grid.column)) {
          columns.put(grid.column, 0);
        }
      }

      // Receives and processes lines from Master
      while (true) {
        MPI.COMM_WORLD.Recv(recvObjectArray, 0, 1, MPI.OBJECT, master, tag);
        line = (String) recvObjectArray[0];

        // Stop receiving
        if (line.equals("terminate")) {
          break;
        } else {
          Integer temp = 0;
          String[] words = line.split(":");

          // Searches for point coordinates in each line
          for (int i = 0; i < words.length; i ++) {
            if (words[i].equals("\"Point\",\"coordinates\"")) {
              String[] latlon = words[i + 1].replaceAll("]}}},", "").replace("[", "").split(",");
              try {
                Float x = Float.parseFloat(latlon[1]);
                Float y = Float.parseFloat(latlon[0]);

                // Checks if points are in a grid and increment counters
                for (Grid grid: grids) {
                  if (grid.inGrid(x, y)) {
                    temp = posts.get(grid.name);
                    posts.remove(grid.name);
                    posts.put(grid.name, temp + 1);

                    temp = rows.get(grid.row);
                    rows.remove(grid.row);
                    rows.put(grid.row, temp + 1);

                    temp = columns.get(grid.column);
                    columns.remove(grid.column);
                    columns.put(grid.column, temp + 1);
                    break;
                  }
                }
              } catch (NumberFormatException e) {
                break;
              }
            }
          }
        }
      }
      postsArray[0] = (Object) posts;
      rowsArray[0] = (Object) rows;
      columnsArray[0] = (Object) columns;
    }

    // Ensures processes are synchronised before combining results
    MPI.COMM_WORLD.Barrier();
    if (size > 1) {
      // Reduces HashMap counts
      HashMerge hashMerge = new HashMerge();
      Op op = new Op(hashMerge, true);
      MPI.COMM_WORLD.Reduce(postsArray, 0, postsArray, 0, 1, MPI.OBJECT, op, master);
      MPI.COMM_WORLD.Reduce(rowsArray, 0, rowsArray, 0, 1, MPI.OBJECT, op, master);
      MPI.COMM_WORLD.Reduce(columnsArray, 0, columnsArray, 0, 1, MPI.OBJECT, op, master);
    }

    // Sorts HashMaps in decreasing order
    List<Map.Entry<String, Integer>> sortedPosts = new ArrayList<>(posts.entrySet());
    Collections.sort(sortedPosts, new Comparator<Map.Entry<String, Integer>>() {
      public int compare(Map.Entry<String, Integer> x, Map.Entry<String, Integer> y) {
        return y.getValue().compareTo(x.getValue());
      }
    });
    List<Map.Entry<String, Integer>> sortedRows = new ArrayList<>(rows.entrySet());
    Collections.sort(sortedRows, new Comparator<Map.Entry<String, Integer>>() {
      public int compare(Map.Entry<String, Integer> x, Map.Entry<String, Integer> y) {
        return y.getValue().compareTo(x.getValue());
      }
    });
    List<Map.Entry<String, Integer>> sortedColumns = new ArrayList<>(columns.entrySet());
    Collections.sort(sortedColumns, new Comparator<Map.Entry<String, Integer>>() {
      public int compare(Map.Entry<String, Integer> x, Map.Entry<String, Integer> y) {
        return y.getValue().compareTo(x.getValue());
      }
    });

    // Prints the final output
    if (rank == master) {
      System.out.println("// Rank by unit");
      for (Map.Entry<String, Integer> entry: sortedPosts) {
        System.out.println(entry.getKey() + ": " + entry.getValue() + " posts");
      }
      System.out.println("// Rank by row");
      for (Map.Entry<String, Integer> entry: sortedRows) {
        System.out.println(entry.getKey() + "-Row: " + entry.getValue() + " posts");
      }
      System.out.println("// Rank by column");
      for (Map.Entry<String, Integer> entry: sortedColumns) {
        System.out.println("Column " + entry.getKey() + ": " + entry.getValue() + " posts");
      }
      endTime = System.nanoTime();
      System.out.println("Time: " + ((endTime - startTime) / 1000000000) + "s");
    }

    MPI.Finalize();
  }

}
